main/src/library/Iterator.flix

```diff
@@ -14,179 +14,158 @@
  * limitations under the License.
  */
 
-pub enum Iterator[a: Type, r: Region] {
-    case Iterator(Iterator.Done[r], Iterator.Next[a, r])
+pub enum Iterator[a: Type, ef: Bool, r: Region] {
+    case Iterator(Region[r], Unit -> Iterator.Step[a] \ {r, ef})
 }
 
+/*
 instance Newable[Iterator[a]] {
-    pub def new(r: Region[r]): Iterator[a, r] \ { Read(r), Write(r) } = Iterator.new(r)
-}
-
-instance Scoped[Iterator[a]] {
-    pub def regionOf(_: Iterator[a, r]): Region[r] = unsafe_cast () as Region[r]
+    pub def new(r: Region[r]): Iterator[a, ef, r] \ { Read(r), Write(r) } = Iterator.iterate(r)
 }
 
 ///
 /// Alias for `Foldable.iterator`.
 ///
-pub def iterator(r: Region[r], t: t[a]): Iterator[a, r] \ Write(r) with Foldable[t] =
+pub def iterator(r: Region[r], t: t[a]): Iterator[a, ef, r] \ Write(r) with Foldable[t] =
     Foldable.iterator(r, t)
+*/
 
 namespace Iterator {
 
-    ///
-    /// The type of the done function.
-    ///
-    pub type alias Done[ef: Region] = Unit -> Bool \ ef
+    pub enum Step[a] {
+        case Ans(a)
+        case Skip
+        case Done
+    }
 
     ///
-    /// The type of the next function.
+    /// Returns an iterator built with the stepper function `f`.
     ///
-    pub type alias Next[a: Type, ef: Region] = Unit -> a \ ef
+    pub def iterate(rc: Region[r], f: Unit -> Option[a] \ ef): Iterator[a, ef, r] =
+        let f1 = () -> match f() {
+            case Some(a) => Ans(a)
+            case None    => Done
+        };
+        let iterF = () -> unsafe_cast f1()  as _ \ { ef, r };
```

Comment:
I assume this cast is here because otherwise the type would be: `Iterator[a, ef, Pure]`. Is that right?

If so, my idea from earlier about doing a no-op effect in `rc` might be useful.

We could even consider adding e.g.:

```
def touch(_rc: Region[r]): Unit \ r = unsafe_cast () as Unit \ r
```

to the library-- which could then be used in `f1`.

Reply:
Let me know if I should give more details.

Reply:
I didn't need `touch` in the end though it might still be useful.

I had in mind that one or other of these might be useful but didn't explore them at this time:

~~~

liftRegion(rc: Region[r1], iter: Iterator[a, ef, r]): Iterator[a, ef and r1, r]

liftRegion(rc: Region[r1], f: Unit -> Iterator[a, ef, r] \ r1): Iterator[a, ef and r1, r]

~~~